### PR TITLE
chore: Update development guide to include information on dependency updates

### PR DESCRIPTION
**Pull request by qingchm**
#### Which issue(s) does this change fix?
N.A.

#### Why is this change necessary?
For contributions with , there is currently no instructions or checks on which required files are updated. The is causing confusion to both contributors and maintainers. For contributors they don't have instructions on which files to update, and for maintainers the PRs are harder to review as these misses are hard to be spotted.

#### How does it address the issue?
It adds a dependency update section on the development guide page for the community to reference to. It states what circumstances each file under requirements folder needs to be updated.

#### What side effects does this change have?
N.A.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by mndeveci — APPROVED** [review]

File: DEVELOPMENT_GUIDE.md [hunk 1]
```diff
@@ -328,6 +328,18 @@ conventions are best practices that we have learnt over time.
     comments.
     
     
+### Dependency Updates
+
+Please update all the required files if the changes involve a version update on a dependency or to include a new dependency. The requirements files are located inside the `requirements folder`.
```
Comment (sriram-mv) [inline DEVELOPMENT_GUIDE.md:333]: The contents looks good, could we structure it such there are 3 parts, base, reproducible and installer?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: DEVELOPMENT_GUIDE.md (hunk 1 above)
Comment (qingchm) [reply DEVELOPMENT_GUIDE.md:333]: Yep makes sense! I can break down the sections further for this to be clearer

File: DEVELOPMENT_GUIDE.md [hunk 2]
```diff
@@ -328,6 +328,27 @@ conventions are best practices that we have learnt over time.
...
+
+#### base.txt for SAM CLI code dependencies
+For dependencies used in SAM CLI code, update `base.txt` in `requirements` folder. To update `base.txt` file, simply follow the current convention and input the dependency name plus version, together with any necessary comment. For more information on the operators to be used for restricting compatible versions, read on [python's enhancement proposals](https://peps.python.org/pep-0440/#compatible-release).
+
+#### reproducible-linux.txt for SAM CLI code dependencies
+For dependencies used in SAM CLI code, also remember to update`reproducible-linux.txt` in `requirements` folder and `THIRD-PART-LICENSESD` in `installer/assets` folder. To update the `reproducible-linux.txt`, run the following script to replace the file:
```
Comment (sriram-mv) [inline DEVELOPMENT_GUIDE.md:339]: `THIRD-PART-LICENSESD` to `THIRD-PARTY-LICENSES`

File: DEVELOPMENT_GUIDE.md [hunk 3]
```diff
@@ -328,6 +328,27 @@ conventions are best practices that we have learnt over time.
...
+```
+make update-reproducible-reqs
+```
+Note that this is a fully auto-generated file, any manual changes to reproducible-linux.txt will not last after the next update running the above script. As for updating the `THIRD-PARTY-LICENSE`, find the corresponding dependency entry in the license file (usually grouped by licensing organization) and update the versions. For adding a new dependency, look up for its licensing organization through PyPi and update the corresponding section. If the license is from GNU or another license type not included in the file, please contact the repository maintainers first. If you are not familiar with working with this file, please contact one of the repository maintainers or cut an issue to help with the update.
```
Comment (sriram-mv) [inline DEVELOPMENT_GUIDE.md:343]: same here.

File: DEVELOPMENT_GUIDE.md [hunk 4]
```diff
@@ -328,6 +328,27 @@ conventions are best practices that we have learnt over time.
...
+
+#### dev.txt for SAM CLI test dependencies
+For changing dependencies used for `make pr` checks and test related dependencies, update `dev.txt` in `requirements` folder only.
+
+#### pyinstaller-build.txt for SAM CLI native installer build dependencies
+For changing dependencies needed to run PyInstaller builds (to run `build-mac.sh` or `build-linux.sh` in `installer/pyinstaller` folder), modify `pyinstaller-build.txt`.
```
Comment (sriram-mv) [inline DEVELOPMENT_GUIDE.md:349]: For changing python dependencies needed for creating builds through Pyinstaller? does that sound better?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: DEVELOPMENT_GUIDE.md (hunk 4 above)
Comment (qingchm) [reply DEVELOPMENT_GUIDE.md:349]: Sounds good, let me re-word this a bit!

File: DEVELOPMENT_GUIDE.md [hunk 5]
```diff
@@ -328,6 +328,27 @@ conventions are best practices that we have learnt over time.
...
     
+### Dependency Updates
+
+Please update all the required files if the changes involve a version update on a dependency or to include a new dependency. The requirements files are located inside the `requirements folder`.
+
+#### base.txt for SAM CLI code dependencies
+For dependencies used in SAM CLI code, update `base.txt` in `requirements` folder. To update `base.txt` file, simply follow the current convention and input the dependency name plus version, together with any necessary comment. For more information on the operators to be used for restricting compatible versions, read on [python's enhancement proposals](https://peps.python.org/pep-0440/#compatible-release).
+
+#### reproducible-linux.txt for SAM CLI code dependencies
+For dependencies used in SAM CLI code, also remember to update`reproducible-linux.txt` in `requirements` folder and `THIRD-PART-LICENSES` in `installer/assets` folder. To update the `reproducible-linux.txt`, run the following script to replace the file:
```
Comment (sriram-mv) [inline DEVELOPMENT_GUIDE.md:339]: nit: typo on 3rd party licenses.

**Review by sriram-mv — APPROVED** [review]

**Pull request merged**